contracts/tokens/SecurityToken.sol

```diff
@@ -616,85 +619,65 @@ contract SecurityToken is StandardToken, DetailedERC20, ReentrancyGuard, Registr
-            return false;
-        }
-
-        for (uint8 i = 0; i < modules[PERMISSIONMANAGER_KEY].length; i++) {
-            if (IPermissionManager(modules[PERMISSIONMANAGER_KEY][i]).checkPermission(_delegate, _module, _perm)) {
-                return true;
-            }
-        }
+    function _burn(address _from, uint256 _value) internal returns (bool) {
+        require(_value <= balances[_from], "Value too high");
+        require(_updateTransfer(_from, address(0), _value), "Burn is not valid");
+        balances[_from] = balances[_from].sub(_value);
+        totalSupply_ = totalSupply_.sub(_value);
+        emit Burnt(_from, _value);
+        emit Transfer(_from, address(0), _value);
+        return true;
     }
 
     /**
-     * @notice used to set the token Burner address. It only be called by the owner
-     * @param _tokenBurner Address of the token burner contract
+     * @notice Burn function used to burn the securityToken
+     * @param _value No. of tokens that get burned
      */
-    function setTokenBurner(address _tokenBurner) external onlyOwner {
-        tokenBurner = ITokenBurner(_tokenBurner);
+    function burn(uint256 _value) checkGranularity(_value) onlyModule(BURN_KEY) public returns (bool) {
+        require(_burn(msg.sender, _value));
+        return true;
     }
 
     /**
-     * @notice Burn function used to burn the securityToken
-     * @param _value No. of token that get burned
+     * @notice Burn function used to burn the securityToken on behalf of someone else
+     * @param _from Address for whom to burn tokens
+     * @param _value No. of tokens that get burned
      */
-    function burn(uint256 _value) checkGranularity(_value) public returns (bool) {
-        _adjustInvestorCount(msg.sender, address(0), _value);
-        require(tokenBurner != address(0), "Token Burner contract address is not set yet");
-        require(_verifyTransfer(msg.sender, address(0), _value, true), "Transfer is not valid");
-        require(_value <= balances[msg.sender], "Value should no be greater than the balance of msg.sender");
-        _adjustBalanceCheckpoints(msg.sender);
-        _adjustTotalSupplyCheckpoints();
-        // no need to require value <= totalSupply, since that would imply the
-        // sender's balance is greater than the totalSupply, which *should* be an assertion failure
-
-        balances[msg.sender] = balances[msg.sender].sub(_value);
-        require(tokenBurner.burn(msg.sender, _value), "Token burner process is not validated");
-        totalSupply_ = totalSupply_.sub(_value);
-        emit Burnt(msg.sender, _value);
-        emit Transfer(msg.sender, address(0), _value);
+    function burnFrom(address _from, uint256 _value) checkGranularity(_value) onlyModule(BURN_KEY) public returns (bool) {
+        require(_value <= allowed[_from][msg.sender]);
+        require(_burn(_from, _value));
+        allowed[_from][msg.sender] = allowed[_from][msg.sender].sub(_value);
```

Comment:
I probably think we should deduct the allowed balance first and then call `_burn()` function. I am not 100 % sure but the code seems vulnerable to cross-function race conditions (`transferFrom()` and `burnFrom()`).